### PR TITLE
[Roll Forward] Enable Python 3 for Bazel to Run Tests

### PR DESCRIPTION
**Pull request by lidizheng**
The original PR #17644 affected internal tooling because of the Bazel `sys.path` to `site-directory` hack. I suspect it is because the internal tooling has extremely huge number of `sys.path` (considering monorepo) which results in consuming too many resources, which eventually cause the hang while interop testing.

To prevent the hack from corrupting other environment, this PR adds a check to determine whether the runtime is Bazel or not.

### CODE REVIEW

**Comment by lidizheng** [comment]
Related issue: #17622

**Comment by lidizheng** [comment]
@ericgribkoff @gnossen PTAL. This PR is adding a try/catch clause around the invocation of the Bazel specific hack function, so in other build environment, the Bazel-hack package will not be existed, hence will not be affected. I cherrypicked this PR to mono-repo, and it passed.

File: src/python/grpcio_tests/tests/reflection/_reflection_servicer_test.py [hunk 1]
```diff
@@ -50,6 +50,16 @@ def _file_descriptor_to_proto(descriptor):
 
 class ReflectionServicerTest(unittest.TestCase):
 
+    # TODO(https://github.com/grpc/grpc/issues/17844)
+    # Bazel + Python 3 will result in creating two different instance of
+    # DESCRIPTOR for each message. So, the equal comparison between protobuf
+    # returned by stub and manually crafted protobuf will always fail.
+    def _assert_sequence_of_proto_equal(self, x, y):
+        self.assertSequenceEqual(
+            list(map(lambda x: x.SerializeToString(), x)),
```
Comment (gnossen) [inline src/python/grpcio_tests/tests/reflection/_reflection_servicer_test.py]: I'd suggest `tuple(proto.SerializeToString() for proto in x)`

File: third_party/py/python_configure.bzl [hunk 2]
```diff
@@ -138,10 +138,13 @@ def _symlink_genrule_for_dir(repository_ctx,
 
 def _get_python_bin(repository_ctx):
     """Gets the python bin path."""
-    python_bin = repository_ctx.os.environ.get(_PYTHON_BIN_PATH)
-    if python_bin != None:
-        return python_bin
-    python_bin_path = repository_ctx.which("python")
+    python_bin = repository_ctx.os.environ.get(_PYTHON_BIN_PATH, 'python')
+    if not '/' in python_bin and not '\\' in python_bin:
```
Comment (gnossen) [inline third_party/py/python_configure.bzl]: Maybe `not os.path.isabs(python_bin)`?

Comment (lidizheng) [reply third_party/py/python_configure.bzl]: I wasn't sure if user will only pass absolute path to `PYTHON_BIN_PATH`. In our `tools` shell files, the environment variable `PYTHON` often set to relative path to point to the virtual environment Python binary.

File: src/python/grpcio_tests/tests/reflection/_reflection_servicer_test.py (hunk 1 above)
Comment (lidizheng) [reply src/python/grpcio_tests/tests/reflection/_reflection_servicer_test.py]: That's a good idea. TIL.

File: third_party/py/python_configure.bzl (hunk 2 above)
Comment (gnossen) [reply third_party/py/python_configure.bzl]: Understood. My point is that we should avoid writing special cases for different platforms wherever possible.

Comment (lidizheng) [reply third_party/py/python_configure.bzl]: I don't like to check it in this way either... Do you have better solution than the two slash checks?

Comment (lidizheng) [reply third_party/py/python_configure.bzl]: Updated to `if not repository_ctx.path(python_bin).exists:`, seems this is the way Skylark (Starlark) check file existence.

**Pull request merged**